### DISCUSSION TITLE
Component type switching via LUA doesn't update properly (uiFixedSlider to uiSlider)

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#262] [Q&A] [answered]
Hello,
I struggle to find a way to "refresh" a modulator that has changed its type from uiFixedSlider to uiSlider

I am using a bank of 32 modulators that are assigned values on reception of a sysex message.
MainFxParam01 can be a ReverbDelay Time for an algo, and for another algo MainFxParam01 can be a Chorus Depth, it depends.

Everything is working beautifully but for one thing, if I change the type from uiFixedSlider to uiSlider the modulator locks even if all the other properties are updated perfectly. It doesn't work both ways.

If I save, restart the panel, I get the right modulator values and behaviour etc. this is very strange. 

for exemple :  

MainFxParam01 is working in its uiFixedSlider type :
```
	MainFxParam01 = panel:getModulatorByName("MainFxParam01")
	MainFxParam01:getComponent():setProperty("componentVisibility", 1, false)
	MainFxParam01:getComponent():setProperty("componentVisibleName", "ErType", true)
	MainFxParam01:getComponent():setProperty("uiSliderMin", 0, true)
	MainFxParam01:getComponent():setProperty("uiSliderMax", 1, true)
	MainFxParam01:getComponent():setProperty("uiSliderDoubleClickEnabled", 1, true)
	MainFxParam01:getComponent():setProperty("uiSliderDoubleClickValue", 0, true)
	MainFxParam01:getComponent():setProperty("uiFixedSliderContent","Type-A\nType-B", true)
	MainFxParam01:getComponent():setValue(MainFxParamValue01, true) -- starts at 0
	MainFxParam01:setProperty("modulatorCustomIndex","01", true)
	MainFxParam01:setProperty("modulatorCustomName","", true)
	MainFxParam01:getComponent():setProperty("uiType","uiFixedSlider", true)
```

The modulator type is then switched to uiSlider, 

```
	MainFxParam01LDly = (MainFxParamValue00*256 + MainFxParamValue01)
	console("Displays MainFxParam01LDly : "..(MainFxParam01LDly)) -- MainFxParam01LDly  0.1 to 1480.0ms

	MainFxParam01 = panel:getModulatorByName("MainFxParam01")
	MainFxParam01:getComponent():setProperty("componentVisibility", 1, false)
	MainFxParam01:getComponent():setProperty("componentVisibleName", "L.Dly", true)
	MainFxParam01:getComponent():setProperty("uiSliderMin", 1, true)
	MainFxParam01:getComponent():setProperty("uiSliderMax", 14800, true)
	MainFxParam01:getComponent():setProperty("uiSliderDoubleClickEnabled", 1, true)
	MainFxParam01:getComponent():setProperty("uiSliderDoubleClickValue", 0, true)
	MainFxParam01:getComponent():setValue(MainFxParam01LDly, true)
	MainFxParam01:setProperty("modulatorCustomIndex","01", true)
	MainFxParam01:setProperty("modulatorCustomName","", true)
	MainFxParam01:getComponent():setProperty("uiFixedSliderContent","", true)
	MainFxParam01:getComponent():setProperty("uiType","uiSlider", true)
```

After the switch, I checked in the panel and everything is shown right, the modulator value is ok, the content slider list is reset, the type is ok, ...
But the Modulator is locked, no value is displayed and the indicator is at 0 and cannot be moved.

Did you ever experienced this bug?

I tried to start my script with the uiType definition first on top of the list but the order doesn't matter.

Do you know how to reset the modulator state like a fresh start could do?  Maybe "reconstructing" the modulator with a JUCE command could do the trick, I have no idea how to switch modulator type from fixedslider to slider flawlessly?

Thanks a lot in advance for your help

Damien

**Comment by Possemo** [comment]
Hmm this is an interesting way of doing things but I think it is not intended. You do either insert an uiSlider or an uiFixedSlider. You cannot change it on-the-fly. Btw. why would you want to do that? With a fixedslider you can do everything a slider can do.

**Reply by damiensellier** [reply]
Hi @Possemo I switch from fixed slider to slider because if the "slider content" is not populated with fixed values, the slider cannot work, unless there's a way I didn't figured out yet (finger crossed). At first I thought like you, that a fixed slider could display the current unmapped value but it doesn't. Hence my question. I was wondering, what really happens in ctrlr when you launch the panel that makes the modulator work again that does not happen when I change all those properties via LUA and that I could reproduce while updating. Maybe there's a way to refresh all the modulators like a "repaint" does with the GUI.

**Comment by dobo365** [comment]
As Possemo, I didn't know you could change component type on the fly but one thing is sure is that I would not recommend that.
The reason is that Ctrlr is not 100% stable in terms of properties. 
I'm never switching component types but I got issues with Image resource property appearing twice, I can also see the Tab property not displayed at the same place all the time, etc...
You could foresee enough component of the different types then enable/disable them or hide/unhide them on the fly.
I also don't see the direct need of uiFixedSlider vs uiSlider. I think I used that only once on all my panels but don't remember for what reason.

**Reply by damiensellier** [reply]
@dobo365 Yes that's a nice workaround. I will probably overlay 2 sliders at the same position 1 fixed, the other not, and hide/unhide depending on the task. Great Idea. 

Just to know when the JUCE documentation says to "destruct" a component with tilde prefix ~  : 

`Component::~Component `

if my component is "MainFxParam01" What is the right syntax to achieve the "destruction" with LUA ?
I want to try it in case it resets the component values in ctrlr

[JUCE Component Class Reference](https://docs.juce.com/master/classComponent.html#ae500b88725913a9c7edc6573252eb827)

Thanks

**Comment by damiensellier** [comment, marked as answer]
Alright, everything is fine, tried and tested. 
As @dobo365 suggested I overlayed the 2 different types of sliders (uiSlider, uiFixedSlider). When the parameter needs fixed values I change the componentVisibility of the uiFixedSlider to "1" and hide the uiSlider with "0".

for exemple, on the modulator MainFxParam15 (uiFixedSlider) and its counterpart MainFxParam15s (uiSlider)

```
	panel:getModulatorByName("MainFxParam15"):getComponent():setProperty("componentVisibility", 0, false)  -- hide the fixed slider

	MainFxParam15Hold = (MainFxParamValue42*256 + MainFxParamValue43)
	console("Displays MainFxParam15Hold : "..(MainFxParam15Hold)) -- MainFxParam15Hold 0.1 to +2400.0

	MainFxParam15 = panel:getModulatorByName("MainFxParam15s")
	MainFxParam15:getComponent():setProperty("componentVisibility", 1, false)
	MainFxParam15:getComponent():setProperty("componentVisibleName", "Hold", true)  -- show the uiSlider
	MainFxParam15:getComponent():setProperty("uiSliderMin", 1, true)
	MainFxParam15:getComponent():setProperty("uiSliderMax", 24000, true)
	MainFxParam15:getComponent():setValue(MainFxParam15Hold, true)
	MainFxParam15:setProperty("modulatorCustomIndex","15", true)
	MainFxParam15:setProperty("modulatorCustomName","", true)
	MainFxParam15:getComponent():setProperty("uiType","uiSlider", true)
```

So, for the people who wants to switch the component type, since the type is not updated properly by Ctrlr, this is the way to go.

**Reply by DRBAUS** [reply]
Yes, I have used this way too with a MUTE and UNMUTE label. Press MUTE and the UNMUTE label appears and vice versa. 

Until I found out that I could actually change the "on mouse down" method that is used by a Modulator. So now I just have a label that uses 2 methods depending on the MUTE or UNMUTE status of that label and the contents of the label changes accordingly. 

Somewhat related .... but not.

**Reply by damiensellier** [reply]
It is related! let's call it the DMSP Double Modulator Single Purpose switching method

**Comment by dobo365** [comment]
Another reflection on this one...
As you can see in the properties tab, some property values are greyed out. Meaning they are read-only.
This is of course from an interface perspective but I think it also means that you should not change those properties by code either.
